### PR TITLE
Finish up functionality of Marionette PoC

### PR DESCRIPTION
**Pull request by MCGallaspy**
## Summary

Finishes the functionality of the toy app (up to creating the "edit" functionality for users). Introduces another core idea: a modal service. Additionally now using Handlebars templates except in trivial cases (trivial == the template is totally static). From the commit messages:

```txt
Why would you want a modal service at all?
I think there are a few reasons:
1. Provides a consistent entry point into modal code.
2. A service can make intelligent policies about queueing
   modals (although this one is dumb and does not).
3. Allows a summoned modal to be decoupled from the originating
   object. In particular, a modal's DOM element is *not* the
   child of the originator.

Moreover any "overlay" object, like KA Lite's sidebar, could
in principle be managed in the same way. Providing a programmatic
interface for dealing with these objects make automated testing
a lot less susceptible to race conditions where underlying
objects are obscured by an overlaid object -- in testing you
can simply request that the modal service be turned off.

But this is just a quick and hacky solution. I think integrating
modals into Kolibri's core could require a little more thought.
```
## TODO

- [ ] Have tests been written for the new code?
- [ ] Has documentation been written/updated?
- [X] New dependencies (if any) added to requirements file
- [ ] Add an entry to CHANGELOG.rst

## Reviewer guidance

The key files to look at are still `components.js` and `components_demo.js`. I've removed some pre-existing cruft and added explanatory comments. The key ideas of this proof of concept (which should be fleshed out before actually being adopted):
* Higher-level components like KolibriCrudView instead of very granular components like XTextInput, YTextInput, ZTextInput, etc. Should be spec'd out for maximum scrutiny.
* Templates closely coupled to style guide e.g. through use of css classes. However I think the style guide will take some adaptation to be really useful in this way. For example, Backbone wraps all views in an extra tag, so the style guide's css rules should account for this.
* Use of Handlebars templates. But I'm compiling in the browser -- in practice we would want to compile during the build process.
* Modal service: here I've implemented a modal service, but it's not available for use by other KolibriModules on the front-end. If the idea has merit, then it should be implemented as a KolibriModule so that the entire application has a unified modal service, but that was a little too complicated for my PoC.

Also, it's not very pretty. The dropdown menus in particular are a little buggy.

### CODE REVIEW

**Comment by MCGallaspy** [comment]
I've added some comments to guide anyone wishing to implement the "edit" functionality for users.

Finally, I remembered one other big TODO for this if we choose to adopt it -- figure out virtual DOM integration. The DOM is aggressively re-rendered in this case. Essentially, whenever the underlying data of the app is changed, *every* view is re-rendered. Obviously this is not so desirable for very large apps unless a virtual DOM is in use. (Although in this particular case it's still pretty snappy on my machine.)

**Comment by rtibbles** [comment]
Could you point to where that is happening? I don't see the change events that would result in every view being rerendered, or is it internal to Marionette? If Marionette rerenders an entire collection view when only one of the models in it changes, I would think that is a fairly big black mark against Marionette.

My feeling about Virtual DOM or Incremental DOM integration is that it is something we should do before public release, but (as we may get reasonable performance without it) not absolutely require it for the MVP.

If we do go with Marionette, then we could leave it with its usual rendering engine to start - and then work on getting the MarionetteLabs Handlebars to Incremental DOM compiler integrated.

However, in the meantime, I think the performance will probably be reasonable with jQuery, and we could just do a drop in replacement later by switching out the way we compile our Handlebars templates, and overriding the Marionette render method.

**Comment by MCGallaspy** [comment]
>Could you point to where that is happening? I don't see the change events that would result in every view being rerendered, or is it internal to Marionette? 

It's not internal to Mn. Mn 2.4.5 only implicitly re-renders items in a CollectionView. In all other cases the user is responsible for triggering a re-render. I do so here: https://github.com/MCGallaspy/kolibri/blob/um-modal/kolibri/core/assets/src/components_demo.js#L280

> If Marionette rerenders an entire collection view when only one of the models in it changes, I would think that is a fairly big black mark against Marionette.

I have not observed this to be the case. I haven't looked at the source, but it appears only items in a collection that are changed are re-rendered.

> My feeling about Virtual DOM or Incremental DOM integration is that it is something we should do before public release, but (as we may get reasonable performance without it) not absolutely require it for the MVP.

Maybe. With a larger app that contains many more DOM nodes then my aggressive re-rendering will get pokey. In contrast having each view manage when it or its child views are re-rendered is somehow conceptually simpler but a lot more verbose, and all that comes with verbosity. I think there are some trade offs to be made:

             | Aggressive re-rendering | View-managed re-rendering
----------|-----------------------------|-------------------------------
vDOM now |       High up-front dev time. IMO the ideal solution.           |    Most complexity. Probably not a good use case for vDOM, since view-managed re-rendering and vDOM overlap in the problems they're solving (i.e. selective re-rendering).
vDOM later |               Will likely encounter performance issues before vDOM integration. Integration later may require significant refactoring. Lowest up-front dev time, but sort of a gamble that performance will still be *good enough*.                |                    I'll interpret this as view-managed *for now* and then refactoring to use vDOM *later*. Still has a high up-front dev time, but *maybe* less than my ideal solution. This trades performance now with some complexity (i.e. larger surface area for bugs) for significant refactoring later and less complexity. In other words, taking on technical debt proportional to the # of front-end lines of code before we undertake the refactoring. IMO this approach is short sighted -- let's commit to one paradigm.


I'd rather commit to aggressive + vDOM integration now **or** (my 2nd choice) commit to view-managed re-rendering now and forget about vDOM integration at least for the foreseeable future.

**Comment by rtibbles** [comment]
Taking React as an example, to get good performance, the same kinds of checks that we would want to do with a nested Marionette View implementation would still be applied if we were to use a Virtual DOM.

To improve performance using shouldComponentUpdate (a method that checks whether a change in state should result in a change in the DOM) is highly recommended - https://facebook.github.io/react/docs/component-specs.html#updating-shouldcomponentupdate

What this means is that if we have enough components to make a Marionette aggressive rendering engine slow, we would also have enough to make a virtual DOM slow, due to the large amount of virtual DOM diffing we would have to do, and we would still have to make the kinds of distinctions we would want in a highly nested Marionette implementation.

As such, I think your table presents a false dichotomy - I believe it should read thusly:

             | Aggressive re-rendering | View-managed re-rendering
----------|-----------------------------|-------------------------------
vDOM now |       High up-front dev time. Will give poor performance on anything but very simple apps.           |    Most complexity. A good use case for vDOM (or IDOM), since view-specific re-rendering is required to make vDOM performant.
vDOM later |               Creates large, unwieldy Views with poor separation of concerns. Will encounter performance issues unless app is very simple. Lowest up-front dev time, but sort of a gamble that performance will still be *good enough*, and will leave a messy refactor to divide up the monolithic templates into nested views. Very difficult to unit test precisely and well due to the sprawling area of concern.               |                    Reasonably high up-front dev time, but *probably* less than using a direct VDOM or IDOM implementation due to institutional familiarity with this nested approach in Backbone in KA Lite. This allows performance now with some complexity, but also allows for unit testing of modularized code, meaning that we can be more sure of the exact function of the view code. A transition to a VDOM or IDOM implementation could then be made seamlessly in the future, simply by modifying the way we compile handlebars templates to compile them into an IDOM or VDOM compatible format, and the render method of Marionette to cause them to be applied via an IDOM or VDOM engine.

*TL;DR* nested views are modular, testable, and improve performance whether we use VDOM/IDOM or not. They also do not impinge on our future capacity to implement a VDOM/IDOM engine, precisely because we can avoid doing direct DOM manipulation due to our heavily nested views.

**Comment by MCGallaspy** [comment]
I'm not arguing against nested views -- I'm talking about our rendering strategy. At any rate after spending a couple hours poking through Marionette's rendering logic and taking a look at some vDOM libraries, I'm certainly leaning towards Views that manage their own re-rendering, otherwise IMO we'll be fighting Marionette. If as you say that would actually be something we'd want to do *anway* with a vDOM library (since it's analogous to `shouldComponentUpdate `), all the better.

**Comment by indirectlylit** [comment]
No matter which framework we choose, we'll probably still have some options around how much we want to tune the performance at the cost of introducing additional logic. For example,

* With Backbone+Mn we can re-render the entire DOM tree on every single event, or we can tell certain sub-views to only listen to certain model update events.
* With React we could leave `shouldComponentUpdate` in the default `true` and diff the entire virtual dom, or we can implement `shouldComponentUpdate` to be a bit smarter and only diff certain pieces.

However if we compare the naive version of both, diff-ing the whole virtual DOM is almost guaranteed to be faster re-rendering the whole real DOM.

One other consideration is that re-rendering portions of the DOM can introduce issues when the user is interacting with the elements. One classic example is putting a red validation border on an email input box. If the border is shown by re-rendering the `<input>` element (as opposed to just updating its `class` attribute), the users's cursor position and focus will be lost. Traditionally the way around this would be to use something like JQuery to change the styling, but now the view is no longer stateless and purely declarative. A virtual DOM (or something that behaves similarly) would be able to update the styling without destroying the `<input>` element.

**Comment by indirectlylit** [comment]
Noticed a bug with the 'add user' drop-downs:

* refresh page
* add a new user named 'test'
* 'test' user shows up in drop-down for 'Classroom 2' but not 'Classroom 1'

File: kolibri/core/assets/src/components.js [hunk 1]
```diff
@@ -149,6 +73,132 @@ var CrudCollection = Mn.CollectionView.extend({
...
+            };
+        });
+        this.validators = options.validators;
+        this.valid = true;
+        // Since "template" is a callback, bindAll ensures "this" refers the view instance.
+        _.bindAll(this, 'template');
+    },
+
+    // For event "xyz", you can specify a handler without having to set up a listener using the camel-cased
+    // "onXyz" convention.
```
Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:115]: convenient, but magical

File: kolibri/core/assets/src/components.js [hunk 2]
```diff
@@ -149,6 +73,132 @@ var CrudCollection = Mn.CollectionView.extend({
...
+    // In other words, if the view instance is given a model attribute when instantiated, then the argument
+    // to this function will be a copy of that model's .attributes hash.
+    template: function() {
+        var html =
+            '{{#unless valid}}<span>Invalid Input!</span><br />{{/unless}}' +
+            '{{#each create}}' +
+                '<input data-attr="{{ name }}" type="text" placeholder="{{ name }}" value="{{ default }}" /><br />' +
+            '{{/each}}' +
+            '<button class="create flat-button">Create</button>';
+        return Handlebars.compile(html)({create: this.create, valid: this.valid});
```
Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:88]: I'm finding the name of `this.create` confusing. It sounds like a function, but seems to actually be a list of attribute names.

File: kolibri/core/assets/src/components.js (hunk 1 above)
Comment (rtibbles) [inline kolibri/core/assets/src/components.js:115]: I agree - veering towards a little too much magic. -10 points for Marionette house!

File: kolibri/core/assets/src/components.js (hunk 2 above)
Comment (rtibbles) [inline kolibri/core/assets/src/components.js:88]: Perhaps slightly more verbose property names would help - reserving verbs for method names? So this could become `this.fields_to_create`?

Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:88]: 'to create' might even be implied .. it might simply be called `this.item_fields` or something.

Of course none of this is relevant to Mn in particular. However careful naming can help frame views declaratively vs imperatively. In this case the emphasis would be on _what_ (user-editable item attributes) versus _how_ (creation of DOM elements).

File: kolibri/core/assets/src/components.js (hunk 1 above)
Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:115]: it is nice that it cuts down on boilerplate that would be necessary with vanilla Backbone though

Comment (MCGallaspy) [inline kolibri/core/assets/src/components.js:115]: IMO it's fairly practical -- turns 3 boilerplate tasks of config (creating an events hash, setting up a listener during initialization, creating the callback function) into 1 (creating the callback function). It doesn't "reach" outside of the view.

File: kolibri/core/assets/src/components.js (hunk 2 above)
Comment (MCGallaspy) [inline kolibri/core/assets/src/components.js:88]: Granted, it's an odd name. I just couldn't think up a concise one at the time.

Comment (MCGallaspy) [inline kolibri/core/assets/src/components.js:88]: I wanted to be able to distinguish between fields which should be displayed on a creation form vs fields that should be editable in the parent CrudView's config options -- hence `create` vs the implied `edit`. But perhaps better would be `creatable` & `editable` (or `creatable_fields` & `editable_fields`).

File: kolibri/core/assets/src/components.js [hunk 3]
```diff
@@ -149,6 +73,132 @@ var CrudCollection = Mn.CollectionView.extend({
...
+                    default: self.getAttrVal(old.name)
+                };
+            });
+        }
+    },
+
+    getAttrVal: function(attr) {
+        var el = this.$el.find('input[data-attr=' + attr + ']');
+        return $(el).val();
+    },
```
Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:148]: The handling of `onBeforeRender` and data attributes feels pretty complex. Local view state gets stored in three places: `this.create`, the dom's `data-attr` value, and the dom's `value` attribute at various times.

File: kolibri/core/assets/src/components_demo.js [hunk 4]
```diff
@@ -46,67 +51,11 @@ var ComponentDemoPlugin = KolibriModule.extend({
...
-            this.textInputField.clear();
-        });
-    },
 
-    // This convention is recommended by Marionette for efficient repaints.
-    // See: http://marionettejs.com/docs/v2.4.5/marionette.layoutview.html#efficient-nested-view-structures
-    onBeforeShow: function() {
-        this.showChildView('taglist', this.tagList);
-        this.showChildView('textinput', this.textInputField);
+var User = Backbone.Model.extend({
```
Comment (indirectlylit) [inline kolibri/core/assets/src/components_demo.js:55]: Is there a place that defines the attributes that a `User` can have?

File: kolibri/core/assets/src/components.js (hunk 3 above)
Comment (MCGallaspy) [inline kolibri/core/assets/src/components.js:148]: I see what you mean -- IMO the data attribute stuff is kind of unavoidable. I suspect it's the kind of thing going on behind the scenes for forms in any framework, but you're right that here it's our responsibility to manage the complexity. 

As for onBeforeRender, that's necessitated to keep the form "nice" on rerender, kind of in line with what you had mentioned about input elements losing focus on re-render. In this case, it's so that intermediary values are kept even when validation fails.

The trade-off is whether the global app state should granularly manage the state of each form (i.e. have a disabled attribute for each field, a focus attribute, an attribute representing the value of each input element) vs hiding many of the details by managing some state in the form view itself. Since they're highly interactive, forms seem to experience a "state explosion". Moreover any element that deals with user input I think will face similar issues in terms of reacting to user's actions -- should user input be modeled as a change of global state, or do we introduce local state with the aim of simplifying global state?

File: kolibri/core/assets/src/components_demo.js (hunk 4 above)
Comment (MCGallaspy) [inline kolibri/core/assets/src/components_demo.js:55]: No. But you could define a [validate](http://backbonejs.org/#Model-validate) method on Models and raise exceptions to avoid arbitrary attributes from being set & retrieved.

File: kolibri/core/assets/src/components.js (hunk 3 above)
Comment (indirectlylit) [inline kolibri/core/assets/src/components.js:148]: That's true - many view components need to have local state when that state is irrelevant to the application's global state. Validation state is a common example - I did the same thing by embedding validation state in `tag.validation` in the [riot.js add-user form](https://github.com/indirectlylit/kolibri/blob/riot-toy-app/kolibri/core/assets/src/riot-tags/usermgmt-addform.tag.html).

Here's an [interesting explanation of best practices](https://facebook.github.io/react/docs/interactivity-and-dynamic-uis.html#what-components-should-have-state) regarding React's `props`, which (if I understand correctly) are like options passed in from a parent or global app state, and `state`, which is a view's own internal UI state.

**Comment by MCGallaspy** [comment]
@indirectlylit only noticed your PR a little while ago. Linking here for others to follow: https://github.com/MCGallaspy/kolibri/pull/2
